### PR TITLE
Make memoized functions less annoying to debug

### PR DESCRIPTION
**Pull request by millerdev**
Also clean up memoized function tracebacks.

🐡 Review by commit.

### CODE REVIEW

File: test_memoized.py [hunk 1]
```diff
@@ -1,7 +1,107 @@
...
+def test_memoized_method_get_cache():
+    sub = Subject()
+    eq(sub.mem_func(0), 1)
+    cache = sub.mem_func.get_cache(sub)
+    eq(cache[(0,)], 1)
+
+
+def test_memoized_function_name():
+    @memoized
+    def a_rose(by_any, other_name):
```
Comment (kaapstorm) [inline test_memoized.py]: :grin: sweet.

**Review by kaapstorm — CHANGES_REQUESTED** [review]
I was with you up until the last two commits.

The second-last commit is not too bad, but you are undoing a very recognisable pattern, and the result is that a reader needs to think through why you have written the wrapped inner function the way you have. I'd recommend at least leaving a comment to explain why you have written it like this.

The last commit does a few things that I'd find easier to parse if they were separated out. I don't love this construct:
```python
def foo():
    if do_something():
        return bar, None
    else:
        do_something_else()
        return None, baz
```
That looks suspiciously like three functions hiding inside the same trench coat.

How about something like this?
```python
def memoized(fn):
    def _memoized(*args, **kwargs):
        try:
            value = m.get_cached_value(args, kwargs)
        except KeyError:
            memoize = m.get_memoize(args, kwargs)
            value = fn(*args, **kwargs)
            memoize(value)
        return value

    m = Memoized(fn)
    # _memoized decorated manually so pdb drops in on its `def` line:
    _memoized = functools.wraps(fn)(_memoized)
    _memoized.get_cache = m.get_cache
    _memoized.reset_cache = m.reset_cache
    return _memoized
```

... where the `get_cached_value()` method is broken up, maybe like so ...
```python
    def get_obj_and_key(self, args, kwargs):
        if self.is_method:
            return args[0], self.get_args_tuple(*args, **kwargs)[1:]
        else:
            return None, self.get_args_tuple(*args, **kwargs)

    def get_cached_value(self, args, kwargs):
        """
        Raises KeyError if value is not cached.
        """
        obj, key = self.get_obj_and_key(args, kwargs)
        cache = self.get_cache(obj)
        return cache[key]

    def get_memoize(self, args, kwargs):
        """
        Returns a function that caches the value it is given.
        """
        obj, key = self.get_obj_and_key(args, kwargs)
        cache = self.get_cache(obj)

        def memoize(value):
￼           cache[key] = value

￼        return memoize
```

**Comment by millerdev** [comment]
@kaapstorm what do you think of https://github.com/dimagi/memoized/pull/7/commits/fc9d91ba57d50de869915f7a36acaa498fb6e0f1

I had originally considered this, but didn't do it because it has two more lines to step through with a debugger.

Note: it's important that the function call `fn(*args, **kw)` is not in the `except KeyError` block because that would add the `KeyError` to the context of an exception thrown by the function (see https://github.com/dimagi/memoized/pull/7/commits/792e64bfd2b72a7ef9bb3e32013f5b8ef0f60580), which adds extra cruft to the traceback in Python 3.

**Review by kaapstorm — APPROVED** [review]

**Comment by kaapstorm** [comment]
Nice simplification.

**Review by dannyroberts — APPROVED** [review]
Looks like a nice cleanup! Here's a sort of check list of assumptions I have, but unless there's more to say about one of them, this looks good to merge

- [x] This is an in-place refactor—no changes to the public interface
- [x] This has been tested or has a straightforward way of being tested against commcare-hq
- [x] You have plans to version-bump and push this to pypi

We mostly follow https://hynek.me/articles/sharing-your-labor-of-love-pypi-quick-and-dirty/ for pushing the package. Let me know if you need me to perform a step there (not sure if your credentials are set up for this pypi repo)

**Comment by millerdev** [comment]
@dannyroberts I checked off all the items on your list.

The HQ testing is still pending, but should be straightforward. I plan to push a new version to PyPI and then bump the [required HQ version](https://github.com/dimagi/commcare-hq/blob/825fdddc3aa1f7cc27531605d6f347f89203bd92/requirements/requirements.in#L19) in a PR. If all tests pass then I will consider it a success and merge the HQ PR. If not, then I will fix the issue or revert this PR and push a second version to PyPI.

**Pull request merged**